fcrepo-kernel-impl/src/test/java/org/fcrepo/kernel/impl/cache/UserTypesCacheImplTest.java

```diff
@@ -0,0 +1,141 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fcrepo.kernel.impl.cache;
+
+import static org.apache.jena.graph.NodeFactory.createURI;
+import static org.apache.jena.rdf.model.ModelFactory.createDefaultModel;
+import static org.apache.jena.riot.RDFFormat.NTRIPLES;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+import org.fcrepo.config.FedoraPropsConfig;
+import org.fcrepo.kernel.api.RdfStream;
+import org.fcrepo.kernel.api.identifiers.FedoraId;
+import org.fcrepo.kernel.api.rdf.DefaultRdfStream;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.jena.rdf.model.Model;
+import org.apache.jena.riot.RDFDataMgr;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author pwinckles
+ */
+public class UserTypesCacheImplTest {
+
+    private static final URI ANIMAL_TYPE = URI.create("http://example.org/#Animal");
+    private static final URI CAT_TYPE = URI.create("http://example.org/#Cat");
+
+    private UserTypesCacheImpl cache;
+
+    private FedoraId fedoraId;
+
+    private String sessionId;
+
+    private String sessionId2;
+
+    @Before
+    public void setup() {
+        final var props = new FedoraPropsConfig();
+        props.setUserTypesCacheSize(1024);
+        props.setUserTypesCacheTimeout(10);
+        cache = new UserTypesCacheImpl(props);
+
+        fedoraId = FedoraId.create(UUID.randomUUID().toString());
+        sessionId = UUID.randomUUID().toString();
+        sessionId2 = UUID.randomUUID().toString();
+    }
+
+    @Test
+    public void loadTriplesWhenNotCached() {
+        final var types = cache.getUserTypes(fedoraId, sessionId, rdfSupplier(ANIMAL_TYPE, CAT_TYPE));
+        assertThat(types, containsInAnyOrder(ANIMAL_TYPE, CAT_TYPE));
+    }
+
+    @Test
+    public void loadFromCacheWhenCached() {
+        cache.getUserTypes(fedoraId, sessionId, rdfSupplier(ANIMAL_TYPE, CAT_TYPE));
+
+        final var types = cache.getUserTypes(fedoraId, sessionId, noCallSupplier());
+        assertThat(types, containsInAnyOrder(ANIMAL_TYPE, CAT_TYPE));
+    }
+
+    @Test
+    public void manuallyCachedTypesAreOnlyAvailableInSession() {
+        cache.cacheUserTypes(fedoraId, List.of(ANIMAL_TYPE), sessionId);
+
+        var types = cache.getUserTypes(fedoraId, sessionId, noCallSupplier());
+        assertThat(types, containsInAnyOrder(ANIMAL_TYPE));
+
+        types = cache.getUserTypes(fedoraId, sessionId2, rdfSupplier(ANIMAL_TYPE, CAT_TYPE));
+        assertThat(types, containsInAnyOrder(ANIMAL_TYPE, CAT_TYPE));
```

Comment:
I think I get what you mean here, but just wondering. If you called
```
types = cache.getUserTypes(fedoraId, sessionId2, noCallSupplier());
```
would `types == emptyList()`?

Reply:
No, because the `noCallSupplier()` throws an exception when it's called. It's just used in the tests to ensure that the values are returned from the cache and were not newly loaded.